File: Core/TuningCore/Scala.swift
//
//  Scala.swift
//  Tenney
//
//  Created by Sebastian Suarez-Solis on 10/1/25.
//

import Foundation

/// Scala scale representation (relative to 1/1)
struct ScalaScale: Equatable, Codable {
    var description: String
    /// Each entry is either cents or a ratio string; we store as Ratio for exactness when possible, otherwise as cents
    var entries: [Entry]

    enum Entry: Equatable, Codable {
        case ratio(Ratio)
        case cents(Double)

        var centsValue: Double {
            switch self {
            case .ratio(let r): return r.cents
            case .cents(let c): return c
            }
        }
    }

    // MARK: Parsing

    /// Parse .scl text (forgiving: skips '!' comments, blank lines).
    static func parse(_ text: String) throws -> ScalaScale {
        var lines = text.split(whereSeparator: \.isNewline).map { String($0) }
        // Strip comments and trim
        lines = lines.compactMap { line in
            let t = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !t.isEmpty else { return nil }
            if t.hasPrefix("!") { return nil }
            return t
        }
        guard !lines.isEmpty else { throw NSError(domain: "Scala", code: 1, userInfo: [NSLocalizedDescriptionKey: "Empty .scl"]) }
        let description = lines.removeFirst()
        guard !lines.isEmpty else { throw NSError(domain: "Scala", code: 2, userInfo: [NSLocalizedDescriptionKey: "Missing count line"]) }

        // Count line may include comments; extract first integer
        let countLine = lines.removeFirst()
        let scanner = Scanner(string: countLine)
        var count: Int = 0
        if !scanner.scanInt(&count) { throw NSError(domain: "Scala", code: 3, userInfo: [NSLocalizedDescriptionKey: "Invalid count line"]) }

        var entries: [Entry] = []
        for line in lines {
            if entries.count >= count { break }
            let t = line.split(separator: " ").first.map(String.init) ?? line
            if t.contains("/") {
                // ratio n/d
                let parts = t.split(separator: "/").map { String($0) }
                if parts.count == 2, let n = Int(parts[0]), let d = Int(parts[1]), d != 0 {
                    entries.append(.ratio(Ratio(n,d)))
                } else {
                    // malformed ratio -> try cents
                    if let c = Double(t) { entries.append(.cents(c)) }
                }
            } else if let c = Double(t) {
                entries.append(.cents(c))
            }
        }
        return ScalaScale(description: description, entries: entries)
    }

    /// Serialize to .scl text (cents for generic entries; ratio for exacts).
    func serialize() -> String {
        var out: [String] = []
        out.append("! \(description)")
        out.append("\(entries.count)")
        out.append("!")
        for e in entries {
            switch e {
            case .ratio(let r): out.append("\(r.n)/\(r.d)")
            case .cents(let c): out.append(String(format: "%.5f", c))
            }
        }
        return out.joined(separator: "\n")
    }
}

File: apps/android/formats/src/main/kotlin/com/tenney/formats/Scala.kt
package com.tenney.formats

import kotlin.math.abs
import kotlin.math.log2

data class ScalaScale(
    val description: String,
    val entries: List<Entry>
) {
    sealed class Entry {
        data class Ratio(val ratio: Rational) : Entry()
        data class Cents(val value: Double) : Entry()
    }

    data class Rational(val n: Int, val d: Int) {
        init {
            require(d != 0) { "Denominator cannot be zero" }
        }

        val cents: Double = 1200.0 * log2(n.toDouble() / d.toDouble())

        companion object {
            fun of(n: Int, d: Int): Rational {
                if (d == 0) return Rational(1, 1)
                var nn = n
                var dd = d
                if (dd < 0) {
                    nn = -nn
                    dd = -dd
                }
                val g = gcd(nn, dd)
                return Rational(nn / g, dd / g)
            }

            private fun gcd(a: Int, b: Int): Int {
                var x = abs(a)
                var y = abs(b)
                while (y != 0) {
                    val t = x % y
                    x = y
                    y = t
                }
                return maxOf(1, x)
            }
        }
    }

    companion object {
        fun parse(text: String): ScalaScale {
            var lines = text.split(Regex("\\R")).map { it.trim() }
            lines = lines.filter { it.isNotEmpty() && !it.startsWith("!") }
            require(lines.isNotEmpty()) { "Empty .scl" }
            val description = lines.first()
            lines = lines.drop(1)
            require(lines.isNotEmpty()) { "Missing count line" }
            val countLine = lines.first()
            val countToken = Regex("-?\\d+").find(countLine)?.value
                ?: error("Invalid count line")
            val count = countToken.toInt()
            lines = lines.drop(1)

            val entries = mutableListOf<Entry>()
            for (line in lines) {
                if (entries.size >= count) break
                val token = line.split(Regex("\\s+")).firstOrNull().orEmpty()
                if (token.contains("/")) {
                    val parts = token.split("/")
                    val n = parts.getOrNull(0)?.toIntOrNull()
                    val d = parts.getOrNull(1)?.toIntOrNull()
                    if (n != null && d != null && d != 0) {
                        entries.add(Entry.Ratio(Rational.of(n, d)))
                    } else {
                        token.toDoubleOrNull()?.let { entries.add(Entry.Cents(it)) }
                    }
                } else {
                    token.toDoubleOrNull()?.let { entries.add(Entry.Cents(it)) }
                }
            }

            return ScalaScale(description = description, entries = entries)
        }
    }

    fun serialize(): String {
        val out = mutableListOf<String>()
        out.add("! $description")
        out.add(entries.size.toString())
        out.add("!")
        entries.forEach { entry ->
            when (entry) {
                is Entry.Ratio -> out.add("${entry.ratio.n}/${entry.ratio.d}")
                is Entry.Cents -> out.add(String.format("%.5f", entry.value))
            }
        }
        return out.joinToString("\n")
    }
}
